java/src/main/java/com/microsoft/graph/bulk/impl/ObjectToEdge.java

```diff
@@ -0,0 +1,281 @@
+import com.microsoft.graph.bulk.impl.model.GremlinEdge;
+import com.microsoft.graph.bulk.impl.model.GremlinEdgeVertexInfo;
+import com.microsoft.graph.bulk.impl.model.GremlinPartitionKey;
+import lombok.SneakyThrows;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.apache.commons.lang3.reflect.MethodUtils;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+import static java.lang.reflect.Modifier.isStatic;
+
+public final class ObjectToEdge {
+    private ObjectToEdge() {
+        throw new IllegalStateException("Utility class, should not be constructed");
+    }
+
+    /**
+     * Converts an object defined by a class that has GremlinEdge annotations defined into an instance of a GremlinEdge
+     *
+     * @param from object to convert into a GremlinEdge
+     * @return An instance of the GremlinEdge object based on the values extracted from the object provided
+     */
+    @SneakyThrows
+    public static GremlinEdge toGremlinEdge(Object from) {
+        Class<?> clazz = from.getClass();
+
+        var converted = new GremlinEdge();
+
+        setLabelFromClass(clazz, converted);
+        if (converted.getLabel() == null) {
+            setLabelFromGetter(clazz, converted, from);
+        }
+        setFieldValues(clazz, converted, from);
+
+        return converted;
+    }
+
+
+    /**
+     * Sets the label value from the GremlinEdge class level annotation if present
+     *
+     * @param clazz   the class of the object being used
+     * @param results the GremlinEdge being updated
+     */
+    private static void setLabelFromClass(Class<?> clazz, GremlinEdge results) {
+        var annotationClass = com.microsoft.graph.bulk.impl.annotations.GremlinEdge.class;
+        if (!clazz.isAnnotationPresent(annotationClass)) {
+            throw new IllegalArgumentException(
+                    "Class " + clazz.getSimpleName() + " is missing GremlinEdge annotation");
+        }
+
+        var edgeLabel = clazz.getAnnotation(annotationClass).label();
+
+        if (!edgeLabel.isBlank()) {
+            results.setLabel(edgeLabel);
+        }
+    }
+
+    /**
+     * Sets the label value from the results of calling the method marked with a GremlinLabelGetter annotation
+     *
+     * @param clazz   the class of the object being used
+     * @param results the GremlinEdge being updated
+     * @param from    the instance of the class to extract the values from
+     * @throws InvocationTargetException When the method executed throws an exceptions
+     * @throws IllegalAccessException    Guards in place to prevent exception from being thrown
+     */
+    private static void setLabelFromGetter(Class<?> clazz, GremlinEdge results, Object from)
+            throws InvocationTargetException, IllegalAccessException {
+        for (Method method : MethodUtils.getMethodsWithAnnotation(clazz, GremlinLabelGetter.class)) {
```

Comment:
Currently there is nothing blocking the usage of the annotation more than once. Behavior would be to use the first instance it comes to. 

Can look into adding some logic to prevent more than one of the @GremlinId, @GremlinLable, @GremlinLabelGetter and @GremlinPartitionKey annotations. Each of them shouldn't allow for more than one.

Reply:
Built the annotation validations bits and integrated into the flow right before an object is converted into either a GremlinEdge or GremlinVertex.